services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/QueueAttributesManager.java

```diff
@@ -0,0 +1,153 @@
+    private final AtomicReference<CompletableFuture<Duration>> visibilityTimeoutSecondsFuture = new AtomicReference<>();
+    private final Duration minReceiveWaitTime;
+
+    public QueueAttributesManager(SqsAsyncClient sqsClient, String queueUrl, Duration minReceiveWaitTime) {
+        this.sqsClient = sqsClient;
+        this.queueUrl = queueUrl;
+        this.minReceiveWaitTime = minReceiveWaitTime;
+    }
+
+    /**
+     * Retrieves the received message timeout based on the provided request and queue attributes.
+     *
+     * @param rq The receive message request
+     * @return CompletableFuture with the calculated receive message timeout in milliseconds
+     */
+    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest rq) {
+        CompletableFuture<Duration> waitTimeFuture = defaultWaitTimeSecondsFuture.get();
+
+        if (waitTimeFuture == null) {
+            CompletableFuture<Duration> newWaitTimeFuture = new CompletableFuture<>();
+            if (defaultWaitTimeSecondsFuture.compareAndSet(null, newWaitTimeFuture)) {
+                fetchQueueWaitTime(newWaitTimeFuture);
+            }
+            waitTimeFuture = defaultWaitTimeSecondsFuture.get();
+        }
+
+        return waitTimeFuture.thenApply(waitTime -> calculateWaitTime(rq, waitTime));
+    }
+
+    /**
+     * Fetches the queue wait time from SQS and completes the provided future with the result.
+     *
+     * @param newWaitTimeFuture The future to complete with the fetched wait time
+     */
+    private void fetchQueueWaitTime(CompletableFuture<Duration> newWaitTimeFuture) {
+        GetQueueAttributesRequest request = GetQueueAttributesRequest.builder()
+                                                                     .queueUrl(queueUrl)
+                                                                     .attributeNames(
+                                                                         QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS)
+                                                                     .build();
+        sqsClient.getQueueAttributes(request)
+                 .thenApply(response -> {
+                     String messageWaitTime =
+                         Validate.notNull(response
+                                              .attributes()
+                                              .get(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS),
+                                          QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS +
+                                          " attribute is null in sqs.");
+
+                     return Duration.ofSeconds(Integer.parseInt(messageWaitTime));
+                 })
+                 .thenAccept(newWaitTimeFuture::complete)
+                 .exceptionally(ex -> {
+                     newWaitTimeFuture.completeExceptionally(ex);
+                     return null;
+                 });
```

Comment:
Should we set up future cancellation? If newWaitTimeFuture is cancelled, we should cancel `getQueueAttributes` future